pkg/acpi/rsdp.go

```diff
@@ -0,0 +1,238 @@
+type RSDP struct {
+	sign     [8]byte `Align:"16" Default:"RSDP PTR "`
+	v1CSUM   uint8   // This was the checksum, which we are pretty sure is ignored now.
+	oemid    [6]byte
+	revision uint8  `Default:"2"`
+	obase    uint32 // was RSDT, but you're not supposed to use it any more.
+	length   uint32
+	base     uint64 // XSDT address, the only one you should use
+	checksum uint8
+	_        [3]uint8
+	data     [HeaderLength]byte
+}
+
+var (
+	defaultRSDP = []byte("RSDP PTR U-ROOT\x02")
+	_           = Tabler(&RSDP{})
+)
+
+// Marshal fails to marshal an RSDP.
+func (r *RSDP) Marshal() ([]byte, error) {
+	return nil, fmt.Errorf("Marshal RSDP: not yet")
+}
+
+// NewRSDP returns a new and partially initalized RSDP, setting only
+// the defaultRSDP values, address, length, and signature.
+func NewRSDP(addr uintptr, len uint) []byte {
+	var r [HeaderLength]byte
+	copy(r[:], defaultRSDP)
+	// This is a bit of a cheat. All the fields are 0.
+	// So we get a checksum, set up the
+	// XSDT fields, get the second checksum.
+	r[cSUM1Off] = gencsum(r[:])
+	binary.LittleEndian.PutUint32(r[xSDTLenOff:], uint32(len))
+	binary.LittleEndian.PutUint64(r[xSDTAddrOff:], uint64(addr))
+	r[cSUM2Off] = gencsum(r[:])
+	return r[:]
+}
+
+// Len returns the RSDP length
+func (r *RSDP) Len() uint32 {
+	return uint32(len(r.data))
+}
+
+// AllData returns the RSDP as a []byte
+func (r *RSDP) AllData() []byte {
+	return r.data[:]
+}
+
+// TableData returns the RSDP table data as a []byte
+func (r *RSDP) TableData() []byte {
+	return r.data[36:]
+}
+
+// Sig returns the RSDP signature
+func (r *RSDP) Sig() string {
+	return string(r.data[:8])
+}
+
+// OEMID returns the RSDP OEMID
+func (r *RSDP) OEMID() string {
+	return string(r.data[9:15])
+}
+
+// OEMTableID returns the RSDP OEMTableID
+func (r *RSDP) OEMTableID() string {
+	return "rsdp?"
+}
+
+// Revision returns the RSDP revision, which
+// after 2002 should be >= 2
+func (r *RSDP) Revision() uint8 {
+	return r.revision
+}
+
+// OEMRevision returns the table OEMRevision.
+func (r *RSDP) OEMRevision() uint32 {
+	return 0
+}
+
+// CheckSum returns the table CheckSum.
+func (r *RSDP) CheckSum() uint8 {
+	return uint8(r.checksum)
+}
+
+// CreatorID returns the table CreatorID.
+func (r *RSDP) CreatorID() uint32 {
+	return uint32(0)
+}
+
+// CreatorRevision returns the table CreatorRevision.
+func (r *RSDP) CreatorRevision() uint32 {
+	return 0
+}
+
+// Base returns a base address or the [RX]SDT.
+// It will preferentially return the XSDT, but if that is
+// 0 it will return the RSDT address.
+func (r *RSDP) Base() int64 {
+	Debug("Base %v data len %d", r, len(r.data))
+	b := int64(binary.LittleEndian.Uint32(r.data[16:20]))
+	if b != 0 {
+		return b
+	}
+	return int64(binary.LittleEndian.Uint64(r.data[24:32]))
+}
+
+func readRSDP(base int64) (*RSDP, error) {
+	r := &RSDP{}
+	r.base = uint64(base)
+	for i := range r.data {
+		var d io.Uint8
+		if err := io.Read(base+int64(i), &d); err != nil {
```

Comment:
I'll add a `io.ByteSlice` to make reading multiple bytes easier in the future.

Reply:
thanks